libevm/VMSIMD.cpp

```diff
@@ -0,0 +1,729 @@
+	m_SPP[0] = 0;
+	switch (t)
+	{
+	case 0:
+		for (int i = 0; i < n; ++i)
+		{
+			v8x32(m_SPP[0])[i] = m_code[++m_PC];
+		}
+		break;
+	case 1:
+		for (int i = 0; i < n; ++i)
+		{
+			uint16_t v = m_code[++m_PC];
+			v = (v << 8) | m_code[++m_PC];
+			v16x16(m_SPP[0])[i] = v;
+		}
+		break;
+	case 2:
+		for (int i = 0; i < n; ++i)
+		{
+			uint32_t v = m_code[m_PC];
+			v = (v << 8) | m_code[++m_PC];
+			v = (v << 8) | m_code[++m_PC];
+			v = (v << 8) | m_code[++m_PC];
+			v32x8(m_SPP[0])[i] = v;
+		}
+		break;
+	case 3:
+		for (int i = 0; i < n; ++i)
+		{
+			uint64_t v = m_code[++m_PC];
+			v = (v << 8) | m_code[++m_PC];
+			v = (v << 8) | m_code[++m_PC];
+			v = (v << 8) | m_code[++m_PC];
+			v = (v << 8) | m_code[++m_PC];
+			v = (v << 8) | m_code[++m_PC];
+			v = (v << 8) | m_code[++m_PC];
+			v = (v << 8) | m_code[++m_PC];
+			v64x4(m_SPP[0])[i] = v;
+		}
+		break;
+	default:
+		throwBadInstruction();
+	}
+}
+
+void VM::xget(uint8_t _b, uint8_t _c)
+{
+	// n type t elements in source vector, m type u in get indexes
+	uint8_t const t = elemType(_b);
+	uint8_t const m = nElem(_c);
+	uint8_t const u = elemType(_c);
+	
+	// given the type of the source and index
+	// for every element of the index get the indexed element from the source
+	switch (t)
+	{
+	case 0:
+
+		switch (u)
+		{
+		case 0:
+			for (int i = 0; i < m; ++i)
+				v8x32 (m_SPP[0])[i] = v8x32(m_SP[0])[v8x32 (m_SP[1])[i] % 32];
+			break;
+		case 1:
+			for (int i = 0; i < m; ++i)
+				v16x16(m_SPP[0])[i] = v8x32(m_SP[0])[v16x16(m_SP[1])[i] % 16];
```

Comment:
I see now, XGET really replaces each index in the stack word with a value from source vector, that's why index vector type and result vector type are the same.